rhsm.go

```diff
@@ -229,14 +255,17 @@ func registerActivationKey(orgID string, activationKeys []string, serverURL stri
 		return err
 	}
 
+	options := make(map[string]string)
+	options["environments"] = strings.Join(environments, ",")
+
 	if err := privConn.Object(
 		"com.redhat.RHSM1",
 		"/com/redhat/RHSM1/Register").Call(
 		"com.redhat.RHSM1.Register.RegisterWithActivationKeys",
 		dbus.Flags(0),
 		orgID,
 		activationKeys,
-		map[string]string{},
+		options,
```

Comment:
The registration using activation key & org & content templates does not work.

You cannot do it so simple, when activation key(s) are used for registration. You cannot use environment names as environment IDs. You have to get IDs from environment names in the same way as you do it for registration using username & password.

Reply:
The GetEnvironments() method requires username and password, so the intention here was to assume that the given values are the IDs. 

But now I am wondering if using activation keys with templates should be prevented? The ID is not really usable in the case of content templates. @ptoscano wrote the ticket he want to provide some input.

Reply:
Yeah, you are right. Getting environments requires username and password...

Hmm, I would be easy to simply say that  content templates are supported only with username and password authentication, but user experience would not be the best.

Other approach: requiring environment name for username & password and requiring environment ID (typically some UUID) for activation key authentication would be very very confusing.

Reply:
The problem here is due to the difference in behaviour of two types of "credentials": username & password, activation key(s) (AKs) + org id:
- username & password are real credentials: they are used when authenticating to Candlepin (via HTTP Basic Auth), and thus the identity of the caller is immediately known even before the code of the actual endpoint is run
- AKs are "credentials", yes, however they are actually metadata passed when calling the `POST /consumers` endpoint. They currently cannot be used for anything else than that

Environments have been historically handled in a "special way": so far they are Katello-only feature, and for reasons I'm not even sure about they were implemented in Katello directly (since Katello proxies the communication to its Candlepin instance) as part of AKs. When connecting directly to Candlepin there is (or better, used to be) no possibility to specify environments with AKs. Hence, if the IDs and names of AKs are different, they don't generally matter in a Katello setup.

Strictly speaking, the org ID has the same inconsistency as environments:
- sub-man does resolve the org name from the available ones when registering with username & password, or even what is passed via `--org`
- sub-man uses `--org` as-is when registering with AKs

The above inconsistency for org ID practically has not been an issue because:
- when registering to Hosted, the org ID and names are the same (numeric strings)
- when registering to Katello, the automated registration ways such as the global registration take care of pre-filling the script to run (and the script uses AKs)
- registration to Katello via username & password is not commonly done

Because of the above, IMHO there is not much that can be done to avoid the inconsistencies from Candlepin; my recommendation is to:
- resolve content templates names -> IDs when registering via username & password, as that is usually done manually (and thus the user better expect user-facing bits)
- assume that the specified content templates are IDs when registering via AKs

Reply:
> Because of the above, IMHO there is not much that can be done to avoid the inconsistencies from Candlepin; my recommendation is to:
>
> * resolve content templates names -> IDs when registering via username & password, as that is usually done 
  manually (and thus the user better expect user-facing bits)
> * assume that the specified content templates are IDs when registering via AKs

No, I will never accept anything like this!

Reply:
I have been thinking about this problem and I think that cleanest solution is to use environment IDs in both cases (authentication using username & password, authentication using activation-key & org). Could environment IDs be some e.g. UUID. Yes, it could happen. Do users care? Not really, because if environment names were supported, then they would probably copy paste environment names to command line anyway. So, it does not matter, what they copy paste.

Reply:
A candlepin dev pointed out that candlepin can natively accept the environment name:

```
Candlepin supports using either environment names or ids during registration. Just pass it like this in the consumer payload:
{
  "environments": [
     {
       "name": "my_environment_name"
       }
    ]
 }
```

Any reason to not go that route?

Reply:
It is probably doable, but something like this would be necessary for subscription-manager first: https://github.com/candlepin/subscription-manager/pull/3480